linalg/tensor.hpp

```diff
@@ -0,0 +1,2051 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+/**
+ * @file tensor.hpp
+ *
+ * @brief Implementation of the tensor class
+ */
+
+#ifndef MFEM_INTERNAL_TENSOR_HPP
+#define MFEM_INTERNAL_TENSOR_HPP
+
+#include "dual.hpp"
+#include <type_traits> // for std::false_type
+
+namespace mfem
+{
+namespace internal
+{
+
+#if defined(__CUDACC__)
+#if __CUDAVER__ >= 75000
+#define MFEM_SUPPRESS_NVCC_HOSTDEVICE_WARNING #pragma nv_exec_check_disable
+#else
+#define MFEM_SUPPRESS_NVCC_HOSTDEVICE_WARNING #pragma hd_warning_disable
+#endif
+#else  //__CUDACC__
+#define MFEM_SUPPRESS_NVCC_HOSTDEVICE_WARNING
+#endif
+
+template <typename T, int... n>
+struct tensor;
+
+/// The implementation can be drastically generalized by using concepts of the
+/// c++17 standard.
+
+template < typename T >
+struct tensor<T>
+{
+   using type = T;
+   static constexpr int ndim      = 1;
+   static constexpr int first_dim = 0;
+   MFEM_HOST_DEVICE T& operator[](int /*unused*/) { return values; }
+   MFEM_HOST_DEVICE const T& operator[](int /*unused*/) const { return values; }
+   MFEM_HOST_DEVICE T& operator()(int /*unused*/) { return values; }
+   MFEM_HOST_DEVICE const T& operator()(int /*unused*/) const { return values; }
+   MFEM_HOST_DEVICE operator T() const { return values; }
+   T values;
+};
+
+template < typename T, int n0 >
+struct tensor<T, n0>
+{
+   using type = T;
+   static constexpr int ndim      = 1;
+   static constexpr int first_dim = n0;
+   MFEM_HOST_DEVICE T& operator[](int i) { return values[i]; }
+   MFEM_HOST_DEVICE const T& operator[](int i) const { return values[i]; }
+   MFEM_HOST_DEVICE T& operator()(int i) { return values[i]; }
+   MFEM_HOST_DEVICE const T& operator()(int i) const { return values[i]; }
+   T values[n0];
+};
+
+template < typename T, int n0, int n1 >
+struct tensor<T, n0, n1>
+{
+   using type = T;
+   static constexpr int ndim      = 2;
+   static constexpr int first_dim = n0;
+   MFEM_HOST_DEVICE tensor< T, n1 >& operator[](int i) { return values[i]; }
+   MFEM_HOST_DEVICE const tensor< T, n1 >& operator[](int i) const { return values[i]; }
+   MFEM_HOST_DEVICE tensor< T, n1 >& operator()(int i) { return values[i]; }
+   MFEM_HOST_DEVICE const tensor< T, n1 >& operator()(int i) const { return values[i]; }
+   MFEM_HOST_DEVICE T& operator()(int i, int j) { return values[i][j]; }
+   MFEM_HOST_DEVICE const T& operator()(int i, int j) const { return values[i][j]; }
```

Comment:
For consistency with other objects in MFEM, we may want to switch (in the future) to column-major storage instead of row-major.

Reply:
Another option is to support more general layouts, e.g. allowing arbitrary strides in all tensor dimensions, cf. `tlayout.hpp`.